echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/PubsubEventMonitor.java

```diff
@@ -0,0 +1,125 @@
+import com.netflix.spinnaker.echo.model.trigger.TriggerEvent;
+import com.netflix.spinnaker.echo.pipelinetriggers.PipelineCache;
+import lombok.NonNull;
+import lombok.val;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import rx.Observable;
+import rx.functions.Action1;
+
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Triggers pipelines in _Orca_ when a trigger-enabled pubsub message arrives.
+ */
+@Component
+public class PubsubEventMonitor extends TriggerMonitor {
+
+  public static final String PUBSUB_TRIGGER_TYPE = "pubsub";
+
+  // This seems like a bad idea -- why does this class need its own mapper?
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  private final PipelineCache pipelineCache;
+
+  @Autowired
+  public PubsubEventMonitor(@NonNull PipelineCache pipelineCache,
+                            @NonNull Action1<Pipeline> subscriber,
+                            @NonNull Registry registry) {
+    super(subscriber, registry);
+    this.pipelineCache = pipelineCache;
+  }
+
+  @Override
+  public void processEvent(Event event) {
+    super.validateEvent(event);
+    if (!event.getDetails().getType().equalsIgnoreCase(PubsubEvent.TYPE)) {
+      return;
+    }
+
+    // TODO(jacobkiefer): Translate pubsub message into artifacts via Jinja here.
+    PubsubEvent pubsubEvent = objectMapper.convertValue(event, PubsubEvent.class);
+
+    Observable.just(pubsubEvent)
+        .doOnNext(this::onEchoResponse)
+        .subscribe(triggerEachMatchFrom(pipelineCache.getPipelines()));
+  }
+
+  @Override
+  protected boolean isSuccessfulTriggerEvent(final TriggerEvent event) {
+    PubsubEvent pubsubEvent = (PubsubEvent) event;
+    return pubsubEvent != null;
+  }
+
+  @Override
+  protected Function<Trigger, Pipeline> buildTrigger(Pipeline pipeline, TriggerEvent event) {
+    PubsubEvent pubsubEvent = (PubsubEvent) event;
+    MessageDescription description = pubsubEvent.getContent().getMessageDescription();
+    return trigger -> pipeline
+        .withTrigger(trigger.atMessageDescription(description.getSubscriptionName(), description.getPubsubType().toString()));
+  }
+
+  @Override
+  protected boolean isValidTrigger(final Trigger trigger) {
+    return trigger.isEnabled()
+        && isPubsubTrigger(trigger);
+  }
+
+  @Override
+  protected Predicate<Trigger> matchTriggerFor(final TriggerEvent event) {
+    PubsubEvent pubsubEvent = (PubsubEvent) event;
+    MessageDescription description = pubsubEvent.getContent().getMessageDescription();
+
+    // TODO(jacobkiefer): Need to apply filters specified in triggers here.
+    return trigger -> trigger.getType().equalsIgnoreCase(PUBSUB_TRIGGER_TYPE)
+        && trigger.getPubsubType().equalsIgnoreCase(description.getPubsubType().toString())
+        && trigger.getSubscriptionName().equalsIgnoreCase(description.getSubscriptionName());
```

Comment:
If you have multiple files in one GCS bucket, do each have their own subscription?

Reply:
No, there's one subscription per bucket in the GCS -> Pubsub setup. This is where artifact filters in pipeline triggers come in so you don't trigger on every single change to a bucket. This is obviously not included in this PR -- hopefully arriving shortly.